Add support for geopoly Interface

nikitadol opened the pull request:
Solves https://github.com/simolus3/drift/issues/2892

Questions and suggestions regarding implementation are welcome

More tests and perhaps an example will be added soon

nikitadol committed:
add geopoly support

nikitadol committed:
Merge branch 'simolus3:develop' into develop

nikitadol committed:
`dart format .`

nikitadol committed:
add geopoly to docs

simolus3 commented on drift_dev/lib/src/analysis/results/types.dart:
```diff
@@ -36,7 +36,7 @@ abstract class HasType {
 /// [HasType] is the interface for sql-typed elements and is implemented by
 /// columns.
-class ColumnType {
+sealed class ColumnType {
```
I like the distinction between custom and builtin types here :+1: We should check whether we can avoid the special subclass for geopoly types though.

simolus3 commented on drift/lib/src/runtime/types/mapping.dart:
```diff
@@ -81,6 +81,15 @@ final class SqlTypes {
       return dartValue.rawSqlValue;
     }
 
+    if (dartValue is GeopolyPolygon) {
```
There are a bunch of cases where I didn't do a good job at that, but I'm trying to move the core `package:drift/drift.dart` library to be more agnostic of the underlying DBMS. So supporting geopoly (a sqlite3-only module) here makes it more complicated to make drift support more databases in the future.

I fully support the analysis additions around geopoly types (because these are always opt-in), but I wonder if we should instead implement the runtime support for polygons in a separate library. It can still be part of the drift package, but perhaps under a different import (say `package:drift/extensions/geopoly.dart`, next to things like json1 support). We can then implement geopoly support with [custom types](https://drift.simonbinder.eu/docs/sql-api/types/) (albeit a custom type implemented in the drift package), which also makes it easier to define tables using polygon columns in Dart.
The generator should still apply that custom type by default if return values from the geopoly extension are involved, but that way we can let the core runtime focus on types supported by all SQL implementations.

simolus3 reviewed (commented):
Awesome work with the sqlparser extension! I think we should consider moving the runtime mapping away from the core library to reflect that this is not something all databases support. The rest of this look great!

nikitadol commented:
Regarding all your comments:
I tried doing this via `CustomSqlType`, which I put in `package:drift/extensions/geopoly.dart`, but there are a few problems:
- The only place where I can specify a different type of function columns and arguments is here:
https://github.com/simolus3/drift/blob/78387a361061c49592b2109e42ba3a7fe33c83ff/drift_dev/lib/src/analysis/resolver/drift/sqlparser/mapping.dart#L153-L155
- It is difficult to create a `DartType` (in this exact place) and then pass it to the `CustomColumnType`

I'm still thinking about a solution, I will be glad if you have any suggestions, since `sqlparser` cannot depend on `drift_dev`

Perhaps a mechanism is needed to allow modules to define types

simolus3 commented:
> The only place where I can specify a different type of function columns and arguments is here:

I'm using the `KnownDriftTypes` utility to solve this. `package:drift/src/drift_dev_helper.dart` exports a bunch of classes which we'd like to resolve in `drift_dev`, we can then pick them up from there and turn them into `InterfaceType`s. That library could also export the geopoly types.

> I'm still thinking about a solution, I will be glad if you have any suggestions, since `sqlparser` cannot depend on `drift_dev`

But we only need to inject the actual custom type in `drift_dev`, right? Once the type hint is there it should be preserved by `sqlparser`.

nikitadol committed:
move geopoly to extensions

nikitadol commented:
I had to slightly change the use of `KnownDriftTypes` so as not to change many synchronous methods to asynchronous

Or is this not an option?

simolus3 commented:
> I had to slightly change the use of `KnownDriftTypes` so as not to change many synchronous methods to asynchronous

I think this shouldn't be a problem. We need the known types for almost everything involving the driver, so we might as well preload it.

The failing test worries me a bit as loading more assets may degrade build performance, but since this is all coming from external packages that would invalidate the entire build either way, it shouldn't be a problem.
You can apply this patch to fix the test

```patch
diff --git a/drift_dev/test/backends/build/build_integration_test.dart b/drift_dev/test/backends/build/build_integration_test.dart
index d81f397f..e1828132 100644
--- a/drift_dev/test/backends/build/build_integration_test.dart
+++ b/drift_dev/test/backends/build/build_integration_test.dart
@@ -583,6 +583,8 @@ class MyDatabase {
       return everyElement(
         anyOf(
           isA<AssetId>().having((e) => e.extension, 'extension', '.json'),
+          // Allow reading SDK or other package assets to set up the analyzer.
+          isA<AssetId>().having((e) => e.package, 'package', isNot('a')),
           other,
         ),
       );
```

simolus3 commented on docs/pages/docs/SQL API/extensions.md:
```diff
@@ -78,3 +78,35 @@ The `bm25`, `highlight` and `snippet` functions from fts5 can also be used in cu
+for rendering polygons as [SVG](https://en.wikipedia.org/wiki/SVG), 
+and other similar operations.
+
+To enable the fts5 extension in drift files and compiled queries, modify the
+[build options]({{ "../Generation options/index.md" | pageUrl }}) to include
+`geopoly` in the `sqlite_module` section.
```
Nice :+1: Let's also briefly mention this section here: https://github.com/simolus3/drift/blob/ca0dee4f83234b053039d23b62d9defd007e7686/docs/pages/docs/Generation%20options/index.md?plain=1#L174

nikitadol committed:
update docs

nikitadol committed:
fix no backend error

simolus3 committed:
More docs, integration tests

simolus3 committed:
Merge remote-tracking branch 'origin/develop' into geopoly

simolus3 reviewed (approved):
I've added a simple test to `drift` to make sure the type is roughly doing what it's supposed to. This looks good to me now, but I want to merge #2939 first.

nikitadol committed:
Merge remote-tracking branch 'origin/develop' into develop

nikitadol committed:
fix docs

nikitadol commented:
Thanks for helping
I also fixed some tests and added clarification to the documentation, since by default `geopoly` is not in the `sqlite3_flutter_libs` package

simolus3 committed:
Merge branch 'develop' into geopoly

simolus3 committed:
Expand allowlist for assets in test

simolus3 commented:
LGTM, thanks! :+1:
